3rdparty/everest/include/everest/everest.h

```diff
@@ -0,0 +1,228 @@
+/*
+ *  Interface to code from Project Everest
+ *
+ *  Copyright 2016-2018 INRIA and Microsoft Corporation
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of Mbed TLS (https://tls.mbed.org).
+ */
+
+#ifndef MBEDTLS_EVEREST_H
+#define MBEDTLS_EVEREST_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct mbedtls_ecdh_context;
+typedef struct mbedtls_ecdh_context mbedtls_ecdh_context;
+
+struct mbedtls_x25519_context_;
+
+typedef struct {
+    struct mbedtls_x25519_context_ *ctx;
```

Comment:
We need a pointer because we can't `#include mbedtls/ecdh.h`, i.e. we can only have a forward-declaration, not an actual one. The core reason for this is that `*_get_params` depends on the `mbedtls_ecdh_side` type, which is defined in `ecdh.h`, which in turn needs to `#include "everest/everest.h"` for the new context variant. I'd prefer moving the type definitions out of `ecdh.h` into a separate header, but that's up to you.

We don't strictly need `mbedtls_everest*`; I only added that to allow for additional extensions for which we may want to use different Everest-specific context structures. They could also be new extensions to `ecdh.h` directly, of course.

Reply:
I am sorry I don't understand. Why do we need `#include "mbedtls/ecdh.h"` here?

Reply:
Because `mbedtls_ecdh_side` is defined in `ecdh.h`. See `mbedtls_everest_get_params` which currently takes an `int` as the third argument, which is really a cheat, while `mbedtls_x25519_get_params` does it the proper way (but `x25519.h` is not `#include`d in `ecdh.h`). I don't understand your question about `mbedtls_ecdh_context`; where do we pass that?

Reply:
I think that the cleanest way to resolve this would be to introduce an `everest_ecdh_side` type and translate between that and `mbedtls_ecdh_side` when we dispatch the call in `ecdh.c`.

The `mbedtls_everest_xxx()` functions take an `mbedtls_ecdh_context` pointer as an argument. I thought that the separation would be cleaner if the functions would take `mbedtls_ecdh_context_everest` directly as a parameter (and maybe some extra parameters where it is necessary) and all the code that touches an `mbedtls_ecdh_context` would be in `ecdh.c`.